### PR TITLE
[glean] Add General API...

### PR DESCRIPTION
**Pull request by mdboom**
This adds the top-level API for glean telemetry -- basically a singleton to handle all of the global functionality.

### CODE REVIEW

File: components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+    fun `disabling metrics should record nothing`() {
+        val stringMetric = StringMetricType(
+                disabled = false,
+                category = "telemetry",
+                lifetime = Lifetime.Application,
+                name = "string_metric",
+                sendInPings = listOf("store1")
+        )
+        Glean.setMetricsEnabled(false)
+        try {
```
[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt]: Because we need to reset the global state that changes here (`setMetricsEnabled`).  Otherwise other unrelated tests start failing randomly (depending on order of test run).

File: components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+     *
+     *   - background: When the application goes to the background.
+     *       This triggers sending of the default ping.
+     *   - default: Event that triggers the default pings.
+     *
+     * @param name The name of the event.
+     */
+    fun handleEvent(name: String) {
+        if (name == "background") {
+            sendPing("baseline", "/baseline")
```
[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: I've updated this to store the UUID in `ping_info/ping_id` for now, and updated this PR to WIP until to wait to hear from @mreid-moz on how best to handle that.

File: components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+                    "Metrics should not be recorded if glean is disabled",
+                    StringsStorageEngine.getSnapshot(storeName = "store1", clearStore = false)
+            )
+        } finally {
+            Glean.setMetricsEnabled(true)
+        }
+    }
+
+    @Test
+    fun `disabling event metrics should record only when enabled`() {
```
[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt]: Done.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt [hunk 4]
```diff
@@ -0,0 +1,137 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```
Comment (Dexterp37) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: I think we're missing test coverage for this engine :)

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+
+    /**
+     * Record an experiment as active.
+     *
+     * @param experimentId the id of the experiment
+     * @param branch the active branch of the experiment
+     * @param extra an optional, user defined String to String map used to provide richer experiment
+     *              context if needed
+     */
+    fun setExperimentActive(
```
Comment (Dexterp37) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: We should have test coverage for this function, testing truncation and edge cases, such us what happens if we try to record the same experiment id twice.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+    }
+
+    /**
+     * Set an active experiment as inactive.
+     *
+     * Warns to the logger (but does not raise) if the experiment is not already active.
+     *
+     * @param experimentId the id of the experiment to set as inactive
+     */
+    fun setExperimentInactive(experimentId: String) {
```
Comment (Dexterp37) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: We should have test coverage for this and also check on edge cases, such as removing a non-existing experiment.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+     * Retrieves the data about active experiments
+     *
+     * @return the map of experiment ids to [RecordedExperimentData]
+     */
+    @Synchronized
+    fun getSnapshot(): Map<String, RecordedExperimentData> {
+        return experiments
+    }
+
+    private fun convertExtra(extra: Map<String, String>): JSONObject {
```
Comment (Dexterp37) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: Let's have some documentation for this function. Maybe we can also add direct test coverage for it, using `@VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)` (if you see fit).

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+     *
+     * @param storeName the name of the desired store.  Since experiments are
+     *   included in all stores, this parameter is effectively ignored, but
+     *   must be provided to comply with the StorageEngine interface.
+     * @param clearStore whether or not to clearStore the requested store.
+     *   This parameter is ignored.
+     *
+     * @return the JSONArray containing information about the active experiments
+     */
+    override fun getSnapshotAsJSON(storeName: String, clearStore: Boolean): Any? {
```
Comment (Dexterp37) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: Let's also have test coverage for this :) You can explicitly mark `storeName` as unused by prefixing it with ` @Suppress("UNUSED_PARAMETER")`, so the signature would become 

`override fun getSnapshotAsJSON(@Suppress("UNUSED_PARAMETER") storeName: String, ...`

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt [hunk 9]
```diff
@@ -0,0 +1,137 @@
...
+    /**
+     * Get a snapshot of the stored data as a JSON object.
+     *
+     * @param storeName the name of the desired store.  Since experiments are
+     *   included in all stores, this parameter is effectively ignored, but
+     *   must be provided to comply with the StorageEngine interface.
+     * @param clearStore whether or not to clearStore the requested store.
+     *   This parameter is ignored.
+     *
+     * @return the JSONArray containing information about the active experiments
```
Comment (Dexterp37) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: Isn't this returning a `JSONObject` rather than a `JSONArray`? For such nested types, it would be better to have an example in the docs.

File: components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt]: Nothing is throwing on purpose.  What I'm indicating is that no matter what happens in this test function, the global state needs to be reset at the end (in the `finally` block).  The behavior that this prevents is a failure in this test leading to a failure in other tests simply because they come after this one.

If there is a way to set a `@Before` on every test in the test suite, that would be a viable alternative.  But I think `@Before` just affects tests in the same fixture or file.

File: components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+     * Indicate that an experiment is no longer running.
+     *
+     * @param experimentId The id of the experiment to deactivate.
+     */
+    fun setExperimentInactive(experimentId: String) {
+        ExperimentsStorageEngine.setExperimentInactive(experimentId)
+    }
+
+    private fun sendPing(store: String, docType: String) {
+        val uuid = UUID.randomUUID()
```
[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: And I guess that implies `pingMaker.collect` should return a pair of `content, documentId`.

File: components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+import mozilla.components.service.glean.storages.StorageEngineManager
+
+object Glean {
+    internal const val SCHEMA_VERSION = 1
+
+    // Include our singletons of StorageEngineManager and PingMaker
+    private val storageEngineManager = StorageEngineManager()
+    private val pingMaker = PingMaker(storageEngineManager)
+    private lateinit var configuration: Configuration
+    internal lateinit var httpPingUploader: HttpPingUploader
```
[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: Because it needs to be modified for testing purposes.  AFAICT, that's the only way, short of adding a `setter` to make that work.

File: components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt]: It does like like, however, that we could use `@After`, which might be cleaner / more obvious than `try/finally`.  I'll play with that.

File: components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt (hunk 10 above)
[the remark this answers is not in the document]
Comment (Dexterp37) [reply components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: Yes, that would make things simpler.

File: components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt (hunk 11 above)
[the remark this answers is not in the document]
Comment (Dexterp37) [reply components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: Ah, good point! Thanks for clarifying!

File: components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+            eventMetric.record("buttonA", "event3")
+            val snapshot3 = EventsStorageEngine.getSnapshot(storeName = "store1", clearStore = false)
+            assertEquals(2, snapshot3!!.size)
+        } finally {
+            Glean.setMetricsEnabled(true)
+        }
+    }
+
+    @Test
+    fun `test sending of default pings`() {
```
[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt]: This test is a lot smaller now because it's not sending the event ping.  But I think it's still worth testing the path here (as well as more directly) because it's the only place where all the pieces involved that get integrated and you can confirm that the UUID in the path matches the UUID in the ping payload.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt (hunk 4 above)
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: Oops.  Done.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt (hunk 5 above)
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: I've put most of the testing for this in `ExperimentsStorageEngineTests` since that's where the real work is done, and just made sure that these delegation functions are covered by tests.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt (hunk 7 above)
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: I added documentation and improved the name a bit.  It's really just a helper function for `getSnapshotAsJSON`, where `detekt` was complaining about excessive nesting, so I think it's fine to just test it as an implementation detail of `getSnapshotAsJSON`.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt (hunk 9 above)
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: Yes, indeed.

File: components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (mreid-moz) [reply components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: It's needed in the URL path for skipping duplicate submissions. It's not required inside the ping itself (for example, the [core ping](https://github.com/mozilla-services/mozilla-pipeline-schemas/blob/dev/schemas/telemetry/core/core.9.schema.json) doesn't include it inside the document). If you can't think of a reason why it would be useful on the client-side, I'd lean towards leaving it out.

[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: Thanks.  I'll leave it out of the ping content itself then.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt [hunk 13]
```diff
@@ -0,0 +1,150 @@
...
+     * @param clearStore whether or not to clearStore the requested store.
+     *   This parameter is ignored.
+     *
+     * @return the JSONObject containing information about the active experiments
+     */
+    override fun getSnapshotAsJSON(
+        @Suppress("UNUSED_PARAMETER") storeName: String,
+        @Suppress("UNUSED_PARAMETER") clearStore: Boolean
+    ): Any? {
+        return getSnapshot().let { pingExperiments ->
```
Comment (pocmo) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: Android Studio complains that this outer `return` is unreachable code and that seems to be correct since you already return from inside the block in all cases. Since `getSnapshot()` never returns `null` the block may not be needed:

```Kotlin
    override fun getSnapshotAsJSON(
        @Suppress("UNUSED_PARAMETER") storeName: String,
        @Suppress("UNUSED_PARAMETER") clearStore: Boolean
    ): Any? {
        val pingExperiments = getSnapshot()

        if (pingExperiments.count() == 0) {
            return null
        }

      ...
```

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+                    experimentData.put(getExtraAsJSON(value.extra))
+                }
+                experimentsMap.put(key, experimentData)
+            }
+            return experimentsMap
+        }
+    }
+
+    override val sendAsTopLevelField: Boolean
+        get() = true
```
Comment (pocmo) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/ExperimentsStorageEngine.kt]: nit: If this is always true you can also create a backing field for it and set it once:

```Kotlin
override val sendAsTopLevelField: Boolean = true
```

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/StorageEngine.kt [hunk 15]
```diff
@@ -20,4 +20,11 @@ internal interface StorageEngine {
      *         ancestor is [Object], so we need to return [Any].
      */
     fun getSnapshotAsJSON(storeName: String, clearStore: Boolean): Any?
+
+    /**
+     * Indicate whether this storage engine is sent at the top level of the ping
+     * (rather than in the metrics section).
+     */
+    val sendAsTopLevelField: Boolean
+        get() = false
```
Comment (pocmo) [inline components/service/glean/src/main/java/mozilla/components/service/glean/storages/StorageEngine.kt]: See above

File: components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+     *
+     * Valid events are:
+     *
+     *   - background: When the application goes to the background.
+     *       This triggers sending of the default ping.
+     *   - default: Event that triggers the default pings.
+     *
+     * @param name The name of the event.
+     */
+    fun handleEvent(name: String) {
```
[the remark this answers is not in the document]
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/Glean.kt]: It definitely is limited to a small enum for now, but the design document includes a plan for "custom pings", post MVP, which the application would define in `pings.yaml` file.  That's the reason for using something more open-ended here.  But I actually think we should probably handle that as an (enum == CUSTOM, string) pair, so I'll update this to use an enum for now.

File: components/service/glean/src/main/java/mozilla/components/service/glean/storages/StorageEngine.kt (hunk 15 above)
Comment (mdboom) [reply components/service/glean/src/main/java/mozilla/components/service/glean/storages/StorageEngine.kt]: Unfortunately with
```
    val sendAsTopLevelField: Boolean = false
```
it says `Property initializers are not allowed in interfaces`.

Comment (pocmo) [reply components/service/glean/src/main/java/mozilla/components/service/glean/storages/StorageEngine.kt]: Ops, yes, makes sense. This should only work in `ExperimentsStorageEngine`.

**Comment by mdboom** [comment]
@pocmo: I've addressed all of your comments and rebased on master.  It looks like the failing tests maybe are unrelated, though.  Thoughts?

**Comment by Dexterp37** [comment]
Hey @mdboom , looks like in addition to the unrelated problems, this is also failing for legit reasons:

> e: /build/android-components/components/service/glean/src/test/java/mozilla/components/service/glean/GleanTest.kt: (106, 9): Unresolved reference: ExperimentsStorageEngine

**Comment by mdboom** [comment]
Great.  Rebased and squashed!

**Comment by pocmo** [comment]
🚀
